### PR TITLE
[MRG] Interactive annotations.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/3467

I thought that the right way of doing this is to have 'modes' for different behaviors, so you don't have to plot everything again to select bad channels. Now the annotation mode is toggled by pressing 'a'.

Still to do:
- [x] modification,
- [x] deletion,
- [x] (maybe) text edit for description (mpl does not have textboxes before next release),
- [x] ~~nicer visuals,~~
- [x] docs/tests.

screenshot_2016-12-07_13-09-02

Wdyt?

### CODE REVIEW

**Comment by mmagnuski** [comment]
I was waiting for this to happen - happy it's happening now :)
I will tests it today, but I like it being an annotation mode.

BTW - sometimes (touchpad or when annotation will need to span more than one 'screen') instead of dragging it is more convenient and precise to click for start and then shift-click for stop.

**Comment by agramfort** [comment]
very excited about this feature !

**Comment by jaeilepp** [comment]
> BTW - sometimes (touchpad or when annotation will need to span more than one 'screen') instead of dragging it is more convenient and precise to click for start and then shift-click for stop.

I believe the control keys still don't work on MAC. It is always possible to extend the view though. Perhaps the modifying feature will eventually overcome the problem with precision.

**Comment by jaeilepp** [comment]
Now the addition and removal work (surprisingly smoothly IMO) and maybe now would be a good moment to get some suggestions on the visuals and the usability. The annotation mode is toggled by pressing 'a'. Annotations are made by click/dragging with left mouse button, modified by dragging the edge of existing annotation and deleted with a right click. These instructions could be added to the annotation dialog, if we eventually want it to stay. The mode is exited by closing the annotation dialog or pressing 'a'.

**Comment by Eric89GXL** [comment]
@mmagnuski do you have time to test it out?

@hoechenberger that heart emoji makes me think you might be interested in trying it, too :)

**Comment by mmagnuski** [comment]
Yes, I played with it and it works well, I like it :) Just two things:
* when I am still in annotations mode and close the whole figure the annotation figure should be closed automatically too
* when I drag a limit of one of the annotations so that it overlaps with another annotation of the same type - they should be merged (not necessarily in this PR).

Another, loosely related question: is the sluggish behavior of plotting (not only the annotations) a general matplotlib limitation or in some part limitation of the current implementation (partial functions passing around dict etc.)?

**Comment by hoechenberger** [comment]
@Eric89GXL, I'm currently mostly using MNE only on a headless server (with lots of memory and CPU power) via the Jupyter notebook. Many interactive features work if the IPython magic `%matplotlib notebook` is executed. I would assume that the approach with opening a second window etc. wouldn't work in this environment -- I cannot test until monday, though!

**Comment by agramfort** [comment]
would it be possible to annotate at the channel level?

**Comment by jaeilepp** [comment]
> would it be possible to annotate at the channel level?

Perhaps. That would mean changes to the Annotations class. Maybe another PR.

**Comment by mmagnuski** [comment]
@agramfort @jaeilepp Would be nice (some time later) - and along with annotations for epochs it could play nicely with autoreject.

**Comment by mmagnuski** [comment]
BTW - text description (and maybe - choosing color :) ) would also be good for another PR from my point of view.

**Comment by jaeilepp** [comment]
> Another, loosely related question: is the sluggish behavior of plotting (not only the annotations) a general matplotlib limitation or in some part limitation of the current implementation (partial functions passing around dict etc.)?

I believe the raw plotter is quite lean in the sense that it does not do a lot of runtime processing. With the epochs plotter there are some extra checks made, and it's using line collections, so there is a small difference in speed. Mostly the speed is proportional to the amount of data plotted.

**Comment by mmagnuski** [comment]
@jaeilepp ok, thanks! I was just comparing the speed to matlab (but not the eeglab plotter - which at least some time before was very inefficient) and was wondering whether it is matplotlib that slow or maybe something can be improved (anyone checked the gr backend btw or gr itself?).

**Comment by Eric89GXL** [comment]
The style Travis build was still unhappy so I killed all but the 2.7-full one

**Comment by jaeilepp** [comment]
Ok. This is ready for review. The modifying is a bit laggy because it uses the hover listener, but it's still usable on my computer and it only affects the annotation mode. If it's too slow for someone, maybe we can switch it off and do it some other way...

**Comment by mmagnuski** [comment]
No, I was just curious about the speed limitations of matplotlib. Modification has noticable lag, but its rather esthetical thing - it is still convenient.

File: tutorials/plot_artifacts_correction_rejection.py [hunk 1]
```diff
@@ -133,8 +133,8 @@
 raw.plot(events=eog_events)  # To see the annotated segments.
 
 ###############################################################################
-# It is also possible to draw bad segments interactively using ``raw.plot``
-# (see :ref:`tut_viz_raw`).
+# It is also possible to draw bad segments interactively using
+# :func:`raw.plot <mne.viz.plot_raw>` (see :ref:`tut_viz_raw`).
```
Comment (Eric89GXL) [inline tutorials/plot_artifacts_correction_rejection.py]: Or
```
:meth:`raw.plot <mne.io.Raw.plot>`
```
?

File: mne/cov.py [hunk 2]
```diff
@@ -1197,7 +1197,7 @@ def _unpack_epochs(epochs):
 
 
 def _get_ch_whitener(A, pca, ch_type, rank):
-    """"Get whitener params for a set of channels."""
+    """Get whitener params for a set of channels."""
```
Comment (Eric89GXL) [inline mne/cov.py]: FYI these changes are also in #3831

**Comment by jaeilepp** [comment]
The hover event seems to be very buggy on old mpl. On 1.3 it messes up everything and while on 1.4 it works, it occasionally throws an error, so I disabled the modifying from mpl < 1.5.

**Comment by jaeilepp** [comment]
This is ready for review. The text box would be nice, but I have no clear idea how it could be done reasonably with matplotlib.

**Comment by agramfort** [comment]
ok interactivity works like a charm for me.

in terms of usability current code only allows me to mark bad segments.

I may want to annotate some signal features such as alpha burst, sleep
spindles, k-complex but also distinguish between the different type of
artifacts.

do you see a simple way to do this?

**Comment by jaeilepp** [comment]
I believe the CIs will be happy about it now. Now the descriptions are also modifiable. Text boxes are not yet available in mpl, so the dialog now simply listens to key presses and updates the description accordingly. Suggestions/comments about the usability and visuals are welcome.

**Comment by jaeilepp** [comment]
screenshot_2017-01-12_15-26-23

**Comment by Eric89GXL** [comment]
That looks cool. Just from looking, one suggestion: color the text and circle-outlines based on the color that is used in the annotation-plotting window.

Maybe @jasmainak you could looks since you've been doing some annotation-related things lately?

**Comment by jasmainak** [comment]
looks cool ... I'll take a look tomorrow morning

**Comment by jasmainak** [comment]
A couple of usability feedbacks:

1. Why is some stuff in green and others in red? It's not obvious to me. I think it should all be the same color or all should be different colors.
2. If I write the name of the annotation, then make a segment, then I can go back and edit the text again. And now if I make a segment, it's marked with the new text but I see only one item in the pop up menu. This is confusing
3. If I make an annotation and close the raw.plot(). Now I fire up raw.plot() again and use 'a' to start annotation mode. Now, I can't delete the previous annotations. It gets confusing which ones I can delete and which ones I can't ...

Visual feedback:

1. I think "Annotate bad data by dragging on top ... window" would look better if it's not in bold.

Feature request:

1. Can I have the time stamp of the vertical line highlighted? It can be above the "Help" button. This will allow users to make accurate annotations ...

Hope this much feedback is enough for now ;)

**Comment by jaeilepp** [comment]
I think I've mostly addressed @jasmainak comments.

**Comment by jasmainak** [comment]
@jaeilepp more comments:

1. I cannot make segments if the text in the annotations is empty
2. It's a bit unintuitive that the new labels are getting randomly at any position rather than at the top rather than the bottom. See this for example
sdfdfd
3. I think it's not super clear at the moment how to add a new label. You have to check the label and then make a segment to add a new empty label. Maybe just pressing enter will add a new label? And maybe also put it in the help text how to do this ...
4. The first label starts with "BAD_" but the remaining labels start with just "\_". Perhaps make it consistent to be "BAD_"?

**Comment by jaeilepp** [comment]
@jasmainak I changed the dialog a bit. See how it is now.

screenshot_2017-01-18_14-44-35

**Comment by jasmainak** [comment]
@jaeilepp much better!

There is still the problem of the empty label though. You cannot draw an annotation when the label is empty. Perhaps don't allow the user to "add" if you cannot draw an annotation with empty label?

Also, I noticed that you allow overlapping annotations. Is this on purpose? But when they are overlapping, I cannot select the edge of the segment which is below ...

**Comment by jasmainak** [comment]
Pretty neat that if I drag the edges of one annotation on to another annotation of the same type, they merge together ... but again not sure if it's a "feature" or a "bug"

**Comment by jasmainak** [comment]
I had a chat with @jaeilepp IRL and we agreed that handling overlapping annotations was not worth the added code complexity. Good from my end.

Anyone else wants to look? Otherwise it has my +1

**Comment by jasmainak** [comment]
@jaeilepp don't forget to update `whats_new`!

File: mne/annotations.py [hunk 3]
```diff
@@ -84,6 +84,36 @@ def __init__(self, onset, duration, description,
...
+        """Remove an annotation. Operates inplace.
+
+        Parameters
+        ----------
+        idx : int | list of int
+            Index of the annotation to remove.
+        """
+        self.onset = np.delete(self.onset, idx)
+        self.duration = np.delete(self.duration, idx)
+        self.description = np.delete(self.description, idx)
```
Comment (Eric89GXL) [inline mne/annotations.py]: need tests for these?

File: mne/viz/tests/test_raw.py [hunk 4]
```diff
@@ -75,22 +76,10 @@ def test_plot_raw():
...
-        fig.canvas.key_press_event('pageup')
-        fig.canvas.key_press_event('pagedown')
-        fig.canvas.key_press_event('home')
-        fig.canvas.key_press_event('end')
-        fig.canvas.key_press_event('?')
-        fig.canvas.key_press_event('f11')
-        fig.canvas.key_press_event('escape')
+        for key in ['escape', 'down', 'up', 'right', 'left', 'o', '-', '+',
+                    '=', 'pageup', 'pagedown', 'home', 'end', '?', 'f11',
+                    'escape']:
```
Comment (Eric89GXL) [inline mne/viz/tests/test_raw.py]: escape is in there twice, is it intentional?

File: mne/viz/tests/test_raw.py [hunk 5]
```diff
@@ -134,6 +123,57 @@ def test_plot_raw():
         plt.close('all')
 
 
+@requires_version('matplotlib', '1.2')
```
Comment (Eric89GXL) [inline mne/viz/tests/test_raw.py]: our minimum requirement is 1.3 already so you don't need this

File: mne/viz/utils.py [hunk 6]
```diff
@@ -261,10 +263,10 @@ def _get_help_text(params):
...
     text, text2 = list(), list()
 
-    text.append(u'\u2190 : \n')
-    text.append(u'\u2192 : \n')
-    text.append(u'\u2193 : \n')
-    text.append(u'\u2191 : \n')
+    text.append(u'\u2190 : \n')  # left arrow
+    text.append(u'\u2192 : \n')  # right arrow
+    text.append(u'\u2193 : \n')  # down arrow
+    text.append(u'\u2191 : \n')  # up arrow
```
Comment (Eric89GXL) [inline mne/viz/utils.py]: why not actually put the unicode characters in the file?

https://unicode-table.com/en/sets/arrows-symbols/

File: mne/viz/utils.py [hunk 7]
```diff
@@ -612,6 +616,13 @@ def _radio_clicked(label, params):
     params['plot_fun']()
 
 
+def _get_active_radiobutton(radio):
+    """Helper to find out active radio button."""
+    # XXX: In mpl 1.5 you can do: fig.radio.value_selected
```
Comment (Eric89GXL) [inline mne/viz/utils.py]: you can do a `try/except` then here, no?

File: tutorials/plot_visualize_raw.py [hunk 8]
```diff
@@ -38,8 +38,18 @@
 # with parameter ``scalings``. If you don't know the scaling factor for
 # channels, you can automatically set them by passing scalings='auto'. With
 # ``pageup/pagedown`` and ``home/end`` keys you can adjust the amount of data
-# viewed at once. To see all the interactive features, hit ``?`` or click
-# ``help`` in the lower left corner of the browser window.
+# viewed at once.
+#
+# You can enter annotation mode by pressing ``a`` key. In annotation mode you
+# can mark segments of data (and modify existing annotations) with left mouse
```
Comment (Eric89GXL) [inline tutorials/plot_visualize_raw.py]: the left mouse

File: tutorials/plot_visualize_raw.py [hunk 9]
```diff
@@ -38,8 +38,18 @@
...
+# button. You can use the description of any existing annotation or create a
+# new description by typing when annotation dialog is active. Notice that
```
Comment (Eric89GXL) [inline tutorials/plot_visualize_raw.py]: the annotation dialog

File: tutorials/plot_visualize_raw.py [hunk 10]
```diff
@@ -38,8 +38,18 @@
...
+# description starting with keyword ``'bad'`` means that the segment will be
```
Comment (Eric89GXL) [inline tutorials/plot_visualize_raw.py]: the description

Comment (Eric89GXL) [inline tutorials/plot_visualize_raw.py]: the keyword

File: tutorials/plot_visualize_raw.py [hunk 11]
```diff
@@ -38,8 +38,18 @@
...
+# discarded when epoching the data. Existing annotation can be deleted with
```
Comment (Eric89GXL) [inline tutorials/plot_visualize_raw.py]: Existing annotations

File: tutorials/plot_visualize_raw.py [hunk 12]
```diff
@@ -38,8 +38,18 @@
...
+# right mouse button.  Annotation mode is exited by pressing ``a`` again or
```
Comment (Eric89GXL) [inline tutorials/plot_visualize_raw.py]: the right mouse button

File: mne/viz/utils.py (hunk 6 above)
Comment (jaeilepp) [inline mne/viz/utils.py]: Does not work for me.

File: mne/viz/utils.py (hunk 7 above)
Comment (jaeilepp) [inline mne/viz/utils.py]: This returns an index and ``value_selected`` returns the text, so I would have to map it to the index. I think it's cleaner like this.

File: mne/annotations.py (hunk 3 above)
Comment (jaeilepp) [inline mne/annotations.py]: Added to ``test_annotations.py``.

**Comment by jaeilepp** [comment]
After discussion IRL with @agramfort I changed the colors to of the circle to gray and the edgecolor indicates the color of the annotation. Also, now the latest addition to the label list is automatically activated.
ann_dialog

**Comment by jasmainak** [comment]
oh yes! Much much nicer!

**Comment by agramfort** [comment]
try with this: https://github.com/agramfort/mne-python/commit/d73e3e0263f21c579c756177bf2f404ec041f6c2

but for some reason now text of label overlaps with radio button

do you understand why?

**Comment by jaeilepp** [comment]
This is how it looks now
buttons

**Comment by agramfort** [comment]
is it possible to add some space between button and the text?

also can you do:

```
fig = figure_nobar(figsize=(4.5, 2.75 + len(labels) * 0.75))
fig.patch.set_facecolor('white')
```

so figure is less wide and the background is white.

thx @jaeilepp

**Comment by agramfort** [comment]
@jaeilepp I pushed a cosmit directly to your branch

**Comment by jaeilepp** [comment]
radio
LGTM

**Comment by agramfort** [comment]
thx @jaeilepp !

**Pull request merged**